x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -845,16 +845,21 @@ private int readMultiRegions(
                         // nothing to read, skip
                         continue;
                     }
-                    final CacheFileRegion fileRegion = get(cacheKey, length, region);
-                    final long regionStart = getRegionStart(region);
-                    fileRegion.populateAndRead(
-                        mapSubRangeToRegion(rangeToWrite, region),
-                        subRangeToRead,
-                        readerWithOffset(reader, fileRegion, Math.toIntExact(rangeToRead.start() - regionStart)),
-                        writerWithOffset(writer, fileRegion, Math.toIntExact(rangeToWrite.start() - regionStart)),
-                        ioExecutor,
-                        listeners.acquire(i -> bytesRead.updateAndGet(j -> Math.addExact(i, j)))
-                    );
+                    ActionListener<Integer> listener = listeners.acquire(i -> bytesRead.updateAndGet(j -> Math.addExact(i, j)));
+                    try {
+                        final CacheFileRegion fileRegion = get(cacheKey, length, region);
+                        final long regionStart = getRegionStart(region);
+                        fileRegion.populateAndRead(
+                            mapSubRangeToRegion(rangeToWrite, region),
+                            subRangeToRead,
+                            readerWithOffset(reader, fileRegion, Math.toIntExact(rangeToRead.start() - regionStart)),
+                            writerWithOffset(writer, fileRegion, Math.toIntExact(rangeToWrite.start() - regionStart)),
+                            ioExecutor,
+                            listener
+                        );
+                    } catch (Exception e) {
+                        listener.onFailure(e);
```

Comment:
Can we assert that this is `AlreadyClosedException` ? This could hide some nasty nasty bugs otherwise :)

Reply:
Sure, makes sense, I pushed ded1d5d2b8da0635b2cd06e7c342261ebe9e1694